helper/resource/plugin.go

```diff
@@ -430,6 +443,12 @@ func runProviderCommand(ctx context.Context, t testing.T, f func() error, wd *pl
 	// get closed, and we'll hang here.
 	cancel()
 
+	// For legacy providers, call the StopProvider RPC so the StopContext
+	// goroutine is cleaned up properly.
+	for _, legacyProviderServer := range legacyProviderServers {
+		legacyProviderServer.StopProvider(ctx, nil) //nolint:errcheck // does not return errors
```

Comment:
For what it is worth, we could consider calling this RPC on all providers started by the testing framework as a better approximation to normal Terraform executions, but it only makes sense for sdk/framework-based providers if we enable them to implement close/stop logic that gets called during that RPC.